dbms/src/Flash/Mpp/MPPTunnelSetHelper.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Coprocessor/CHBlockChunkCodec.h>
+#include <Flash/Mpp/MPPTunnelSetHelper.h>
+
+namespace DB::MPPTunnelSetHelper
+{
+TrackedMppDataPacketPtr toPacket(Blocks & blocks, const std::vector<tipb::FieldType> & field_types)
+{
+    CHBlockChunkCodec codec;
+    auto codec_stream = codec.newCodecStream(field_types);
```

Comment:
Might move codec_stream to MPPTunnelSet as a member field, and pass the SetHelper as param? No need to call newCodecStream each time.

Reply:
`codec_stream ` cannot be placed in the mpp tunnel set, because it isn't thread-safe.
The cost of recreating `codec_stream` each time is https://github.com/pingcap/tiflash/blob/53862f2b06d3952d74966f90dde79bf7d5189eb5/dbms/src/Flash/Coprocessor/ CHBlockChunkCodec.cpp#L31-L34, the cost is small.
And Then because `CHBlockChunkCodecStream` reallocates memory each time it encodes, there is no performance benefit to reuse. https://github.com/pingcap/tiflash/blob/53862f2b06d3952d74966f90dde79bf7d5189eb5/dbms/src/Flash/Coprocessor/CHBlockChunkCodec.cpp#L39-L43

Reply:
Ok